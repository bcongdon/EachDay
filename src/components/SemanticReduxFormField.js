// Credit to https://stackoverflow.com/a/42422255

import React from 'react'
import PropTypes from 'prop-types'
import { Input } from 'semantic-ui-react'

export default function SemanticReduxFormField ({ input, label, meta: { touched, error, warning }, as: As = Input, ...props }) {
  function handleChange (e, { value }) {
    return input.onChange(value)
  }
  return (
    <div>
      <As {...input} value={input.value} {...props} onChange={handleChange} error={Boolean(touched && error)} />
      {touched && (warning && <span>{warning}</span>)}
    </div>
  )
}

SemanticReduxFormField.propTypes = {
  as: PropTypes.any,
  input: PropTypes.any,
  label: PropTypes.any,
  meta: PropTypes.any
}
