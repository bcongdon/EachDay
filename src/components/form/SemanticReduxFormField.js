// Credit to https://stackoverflow.com/a/42422255

import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { Message } from 'semantic-ui-react'

class SemanticReduxFormField extends Component {
  static propTypes = {
    as: PropTypes.any,
    input: PropTypes.any,
    label: PropTypes.any,
    meta: PropTypes.any
  }

  constructor (props) {
    super(props)
    this.handleChange = this.handleChange.bind(this)
  }

  handleChange (e, { value }) {
    return this.props.input.onChange(value)
  }

  render () {
    const { input, label, meta, ...props } = this.props
    const { touched, error } = meta

    return (
      <div>
        <this.props.as {...input} value={input.value} {...props} onChange={this.handleChange} error={Boolean(touched && error)} />
        {touched && (error && <Message error content={error} />)}
      </div>
    )
  }
}

export default SemanticReduxFormField
