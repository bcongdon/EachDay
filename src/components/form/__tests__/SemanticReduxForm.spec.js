import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'
import { Form, Message } from 'semantic-ui-react'
import SemanticReduxFormField from '../SemanticReduxFormField'

describe('SemanticReduxForm component', () => {
  it('should render correctly as a text area', () => {
    var form = shallow(
      <SemanticReduxFormField
        as={Form.TextArea}
        meta={{touched: false, error: ''}}
        input={{value: 'foo bar'}}
      />
    )
    expect(toJson(form)).toMatchSnapshot()
  })

  it('should not set error if not touched', () => {
    var form = shallow(
      <SemanticReduxFormField
        as={Form.TextArea}
        meta={{touched: false, error: 'Warning!'}}
        input={{value: 'foo bar'}}
      />
    )
    expect(form.find(Message).length).toBe(0)
    expect(toJson(form)).toMatchSnapshot()
  })

  it('should set error if touched and error', () => {
    var form = shallow(
      <SemanticReduxFormField
        as={Form.TextArea}
        meta={{touched: true, error: 'Warning!'}}
        input={{value: 'foo bar'}}
      />
    )
    expect(form.find(Message).length).toBe(1)
    expect(toJson(form)).toMatchSnapshot()
  })
})
