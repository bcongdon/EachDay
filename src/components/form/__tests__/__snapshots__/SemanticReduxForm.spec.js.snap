// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SemanticReduxForm component should not set error if not touched 1`] = `
<div>
  <FormTextArea
    as={[Function]}
    control={[Function]}
    error={false}
    onChange={[Function]}
    value="foo bar"
  />
</div>
`;

exports[`SemanticReduxForm component should render correctly as a text area 1`] = `
<div>
  <FormTextArea
    as={[Function]}
    control={[Function]}
    error={false}
    onChange={[Function]}
    value="foo bar"
  />
</div>
`;

exports[`SemanticReduxForm component should set error if touched and error 1`] = `
<div>
  <FormTextArea
    as={[Function]}
    control={[Function]}
    error={true}
    onChange={[Function]}
    value="foo bar"
  />
  <Message
    content="Warning!"
    error={true}
  />
</div>
`;
